### PR TITLE
FFI test containerfile

### PR DESCRIPTION
**Pull request by Yarboa**
~This PR is dependent with PR #210,~
~Do not merge~

### CODE REVIEW

File: tests/e2e/tools/FFI/ContainerFile.tools [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM quay.io/centos/centos:stream9
+RUN dnf install -y \
+	dnf-plugin-config-manager \
+	epel-release
+RUN dnf config-manager -y --set-enabled crb
+
+RUN dnf install -y \
+        gcc
```
Comment (dougsland) [inline tests/e2e/tools/FFI/ContainerFile.tools:8]: probably good to add some editor there too, like vim, in case we need to edit the sources.

**Comment by dougsland** [comment]
LGTM, @rhatdan PTAL.

File: tests/e2e/tools/FFI/ContainerFile.tools (hunk 1 above)
Comment (Yarboa) [reply tests/e2e/tools/FFI/ContainerFile.tools:8]: sure

**Comment by Yarboa** [comment]
> /cc @dennisbrendel @priyverm

IT is in Draft, sine i would like to run it through tmt, tmt tests should be added

File: tests/e2e/tools/FFI/ContainerFile.tools [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+RUN mkdir -p /root/tests/FFI/src /root/tests/FFI/bin
+
+COPY ./memory/ASIL/20_percent_memory_eat.c /root/tests/FFI/src/
+COPY ./memory/QM/90_percent_memory_eat.c /root/tests/FFI/src/
+COPY ./disk/file-allocate.c /root/tests/FFI/src/
+
+RUN cd /root/tests/FFI/src && gcc -o ../bin/20_percent_memory_eat 20_percent_memory_eat.c
```
Comment (rhatdan) [inline tests/e2e/tools/FFI/ContainerFile.tools:16]: Can we just add a high level Makefile to compile all of these.

Comment (dougsland) [reply tests/e2e/tools/FFI/ContainerFile.tools:16]: Without test, here an example of Makefile:
```
CC = gcc
CFLAGS = -Wall -g

all: 90_percent_memory_eat 20_percent_memory_eat file_allocate

20_percent_memory_eat: 20_percent_memory_eat.c
	$(CC) $(CFLAGS) -o $@ $<

file-allocate: file-allocate.c
	$(CC) $(CFLAGS) -o $@ $<

90_percent_memory_eat: 90_percent_memory_eat.c
	$(CC) $(CFLAGS) -o $@ $<

clean:
	rm -f file-allocate 20_percent_memory_eat 90_percent_memory_eat

.PHONY: all clean
```

Comment (rhatdan) [reply tests/e2e/tools/FFI/ContainerFile.tools:16]: LGTM

Comment (Yarboa) [reply tests/e2e/tools/FFI/ContainerFile.tools]: Sure will add it

File: tests/e2e/tools/FFI/ContainerFile.tools (hunk 1 above)
Comment (Yarboa) [reply tests/e2e/tools/FFI/ContainerFile.tools:8]: It will big fat vm :) 
OK will add it

File: tests/systemd/prepare.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+# shellcheck disable=SC1091
+
+qm_service_file=$(systemctl show -P  FragmentPath qm)
```
Comment (Yarboa) [inline tests/systemd/prepare.sh:4]: @Yarboa please move all the nested QM unit files attributes to plan prepare
Use only the relevant attributes important to to the test.
Do not forget to return unit file to its previous condition

Comment (rhatdan) [reply tests/systemd/prepare.sh:4]: Talking to yourself?

Comment (Yarboa) [reply tests/systemd/prepare.sh:4]: Yep, sorry :)

File: tests/e2e/tools/FFI/ContainerFile.tools (hunk 2 above)
Comment (Yarboa) [reply tests/e2e/tools/FFI/ContainerFile.tools]: @rhatdan Added make file 
@dougsland preder to maintain slim image for now

File: tests/systemd/prepare.sh (hunk 3 above)
Comment (Yarboa) [reply tests/systemd/prepare.sh]: Done

File: tests/e2e/tools/FFI/ContainerFile.tools (hunk 2 above)
Comment (dougsland) [reply tests/e2e/tools/FFI/ContainerFile.tools]: @Yarboa sure, lets make it fly and during the road we can adjust.

**Comment by dougsland** [comment]
@Yarboa if you are ready to review, please switch from `Draft` to `Ready for review`

**Comment by Yarboa** [comment]
> @Yarboa if you are ready to review, please switch from `Draft` to `Ready for review`

@dougsland still working on Quay.io for qm
Once we have it, need another fix for quay url

**Comment by rhatdan** [comment]
@Yarboa is this still active?

**Comment by Yarboa** [comment]
> @Yarboa is this still active?

@rhatdan Yes, can continue working on it, only now.

File: tests/ffi/memory/test.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+# shellcheck disable=SC1091
+
+. ../memory/prepare.sh
+
+prepare_test
+reload_config
+
+# https://quay.io/repository/centos-sig-automotive/ffi-tools:latest
+podman run --replace -d --name ffi-asil localhost/ffi-tools:latest \
```
Comment (Yarboa) [inline tests/ffi/memory/test.sh:11]: Broken container
https://gitlab.com/CentOS/automotive/container-images/-/tree/main/images/ffi-tools

MR is waiting for fix
@odra

Comment (Yarboa) [reply tests/ffi/memory/test.sh]: Done

**Comment by Yarboa** [comment]
> @Yarboa is this still active?

Yes, hope it is the final iteration

**Comment by Yarboa** [comment]
PR is blocked by this one 
#253

**Comment by Yarboa** [comment]
@odra @dougsland @dennisbrendel @priyverm PTAL

File: tests/e2e/tools/README.md [hunk 5]
```diff
@@ -1 +1,5 @@
 The `remove-containers` script will remove all containers (and images) which name start with node* or control from the output of podman ps. Useful to cleanup environment before executing tests and measures.
+
+### Freedom From Interference
+
+Under FFI c binaries build in different [automotive-repo](https://gitlab.com/CentOS/automotive/container-images/-/tree/main/images/ffi-tools)
```
Comment (dougsland) [inline tests/e2e/tools/README.md:5]: not sure if I follow the 'c' here in this sentence.

File: tests/ffi/memory/prepare.sh [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+   cp "${qm_service_file}" "${qm_service_backup}"
+   # Remove 'DropCapability=sys_resource' enable nested container in QM
+   sed -i 's/DropCapability=sys_resource/#DropCapability=sys_resource/' \
+       /etc/containers/systemd/qm.container
+   # Keep definition for the test not to survive reboot
+   #systemctl set-property --runtime qm ManagedOOMMemoryPressure=kill
+}
+
+reload_config() {
+
```
Comment (dougsland) [inline tests/ffi/memory/prepare.sh:20]: no need white spaces.

File: tests/ffi/memory/prepare.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+# shellcheck disable=SC1091
+
+
+prepare_test() {
+
+   qm_service_file=$(systemctl show -P  SourcePath qm)
+
+   #create backup file for qm unit file
+   qm_service_backup=$(mktemp -d -p /tmp)/qm.service
+   cp "${qm_service_file}" "${qm_service_backup}"
```
Comment (dougsland) [inline tests/ffi/memory/prepare.sh:11]: I would check if the command worked via $?  or use a function to reuse the validation, such:
```
ec_cmd() {
    local cmd="$1"
    eval "$cmd"
    local exit_code="$?"

    if [ "$exit_code" -ne 0 ]; then
        echo "Error: Command failed with exit code $exit_code"
        exit "$exit_code"
    fi
}
```

File: tests/ffi/memory/prepare.sh [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+   sed -i 's/DropCapability=sys_resource/#DropCapability=sys_resource/' \
+       /etc/containers/systemd/qm.container
+   # Keep definition for the test not to survive reboot
+   #systemctl set-property --runtime qm ManagedOOMMemoryPressure=kill
+}
+
+reload_config() {
+
+   systemctl daemon-reload
+   systemctl restart qm
```
Comment (dougsland) [inline tests/ffi/memory/prepare.sh:22]: we should validate here if the command exit correctly or not.

File: tests/ffi/memory/test.sh [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash -eux
+
+# shellcheck disable=SC1091
+
+. ../memory/prepare.sh
+
+prepare_test
+reload_config
+
+podman run --rm --replace -d --name ffi-asil \
```
Comment (dougsland) [inline tests/ffi/memory/test.sh:10]: another good example of check the exit code.

File: tests/ffi/memory/test.sh [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+     quay.io/centos-sig-automotive/ffi-tools:latest \
+     ./ASIL/20_percent_memory_eat > /dev/null
+# Copy container image registry to /var/qm/lib/containers
+image_id=$(podman images | grep quay.io/centos-sig-automotive/ffi-tools | awk -F " " '{print $3}')
+mkdir -p /var/qm/lib/containers/registry
```
Comment (dougsland) [inline tests/ffi/memory/test.sh:15]: I would recommend have a global var session and check if dir already exist before creating it.

QM_REGISTRY_DIR="/var/qm/lib/containers/registry"
if [ ! -d "${QM_REGISTRY_DIR}" ]; then
    mkdir -p ${QM_REGISTRY_DIR}"
fi

File: tests/ffi/memory/test.sh [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+podman push  "${image_id}" dir:/var/qm/lib/containers/registry/tools-ffi:latest
+podman exec -it qm /bin/bash -c \
+       "podman run  --replace --name ffi-qm  dir:/var/lib/containers/registry/tools-ffi:latest \
+       ./QM/90_percent_memory_eat > /dev/null"
+
+if [ $? -eq 137 ]; then
+    echo ffi-qm was killed by SIGKILL
+fi
+systemctl status qm --no-pager | grep "qm.service: A process of this unit has been killed"
```
Comment (dougsland) [inline tests/ffi/memory/test.sh:24]: if this command fails, all good?

**Comment by dougsland** [comment]
In general looks good to me, few comments.

File: tests/e2e/tools/README.md (hunk 5 above)
Comment (Yarboa) [reply tests/e2e/tools/README.md:5]: Ack, typo

File: tests/ffi/memory/prepare.sh (hunk 6 above)
Comment (Yarboa) [reply tests/ffi/memory/prepare.sh:20]: Ack

File: tests/ffi/memory/prepare.sh (hunk 7 above)
Comment (Yarboa) [reply tests/ffi/memory/prepare.sh:11]: Ack

File: tests/ffi/memory/prepare.sh (hunk 8 above)
Comment (Yarboa) [reply tests/ffi/memory/prepare.sh:22]: Ack

File: tests/ffi/memory/test.sh (hunk 9 above)
Comment (Yarboa) [reply tests/ffi/memory/test.sh:10]: Ack

File: tests/ffi/memory/test.sh (hunk 10 above)
Comment (Yarboa) [reply tests/ffi/memory/test.sh:15]: Ack

File: tests/ffi/memory/test.sh (hunk 11 above)
Comment (Yarboa) [reply tests/ffi/memory/test.sh:24]: Ack

**Pull request merged**